### PR TITLE
fixes #7080: symbols are no longer being sorted

### PR DESCRIPTION
**Pull request by dmitri-d**

### CODE REVIEW

**Comment by lzap** [comment]
This needs some testing, I will test against foreman discovery image.

**Comment by lzap** [comment]
EL7 http://koji.katello.org/koji/taskinfo?taskID=140465

EL6 http://koji.katello.org/koji/taskinfo?taskID=140467

Unable to start on EL6:

```
[root@proxybug ~]# service foreman-proxy restart
Stopping foreman-proxy: foreman-proxy was not running.     [FAILED]
Starting foreman-proxy: /usr/share/foreman-proxy/lib/smart_proxy.rb:96:in `https_app': /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in `gem_original_require': /usr/share/foreman-proxy/modules/root/root_api.rb:8: syntax error, unexpected '.', expecting kEND (SyntaxError)
        .collect(&:to_s)
         ^
/usr/share/foreman-proxy/modules/root/root_api.rb:9: syntax error, unexpected '.', expecting kEND
        .sort
         ^
        from /usr/lib/ruby/site_ruby/1.8/rubygems/custom_require.rb:31:in `require'
        from (eval):1:in `https_app'
        from /usr/share/foreman-proxy/lib/smart_proxy.rb:96:in `each'
        from /usr/share/foreman-proxy/lib/smart_proxy.rb:96:in `https_app'
        from /usr/lib/ruby/gems/1.8/gems/rack-1.1.0/lib/rack/builder.rb:46:in `instance_eval'
        from /usr/lib/ruby/gems/1.8/gems/rack-1.1.0/lib/rack/builder.rb:46:in `initialize'
        from /usr/share/foreman-proxy/lib/smart_proxy.rb:95:in `new'
        from /usr/share/foreman-proxy/lib/smart_proxy.rb:95:in `https_app'
        from /usr/share/foreman-proxy/lib/smart_proxy.rb:124:in `launch'
        from /usr/share/foreman-proxy/bin/smart-proxy:6
```

**Comment by lzap** [comment]
@witlessbird you need to put 1.8.7 into your RVM, seriously :-)

Works fine on RHEL7...

**Comment by dmitri-d** [comment]
@lzap: fixed the issue and added tests for root_api

**Comment by lzap** [comment]
:+1: 

```
[root@proxybug ~]# rpm -Uvh --force getfile\?taskID\=140574\&name\=foreman-proxy-1.7.0-0.develop.el6.noarch.rpm 
Preparing...                ########################################### [100%]
   1:foreman-proxy          ########################################### [100%]
[root@proxybug ~]# service foreman-proxy restart
Stopping foreman-proxy: foreman-proxy was not running.     [FAILED]
Starting foreman-proxy:                                    [  OK  ]

[root@proxybug ~]# wget https://localhost:8443/facts --no-check-certificate -O -
--2014-08-14 11:31:14--  https://localhost:8443/facts
Resolving localhost... ::1, 127.0.0.1
Connecting to localhost|::1|:8443... failed: Connection refused.
Connecting to localhost|127.0.0.1|:8443... connected.
WARNING: cannot verify localhost’s certificate, issued by “/CN=Puppet CA: proxybug.zzz.lan”:
  Self-signed certificate encountered.
WARNING: certificate common name “proxybug.zzz.lan” doesn’t match requested host name “localhost”.
HTTP request sent, awaiting response... 200 OK
Length: 2462 (2.4K) [application/json]
Saving to: “STDOUT”

 0% [                                                                                      ] 0           --.-K/s              {"netmask_eth0":"255.255.255.0","rubysitedir":"/usr/lib/ruby/site_ruby/1.8","mtu_eth0":"1500","uptime_days":0,"uptime":"0:06 hours","sshdsakey":"AAAAB3NzaC1kc3MAAACBALlUXvvO+FN1RAOc18ApxiXOxkXaLbluJNFX+By5yJBCgbC2/gDK8qM5oX7mI9uR77LXTRNegEp5QbBCv9jG0DR6QhRZx1bX4uBlwUUgqON9g9lFNTvu4/9h48BEu3Aw+Vu2GdsAXe3FjLI3Iknq0NsSFdgQ6BkWVsqBUOws5UKpAAAAFQDWHJ+RO1vQDZY8OsQJjcSrLHN6wQAAAIBX99+boAQvj59XiBJqsiYD1ece71BattLkMkWDJ/sXbi4p0zjaE3W3NEKciX2PeQjNWi6ZMFlQRE9jhhC7CsO9NSZO1UOotIY8SneYE7Kvk3W+be+JI+pOzzyPYc3U/QaBYMkH/oRe8/7S0PoQgoWqYZ77QLWBVTMxhoR550uOLwAAAIEApG5oIKEwlBAH9P7A7xcBmGYFuVnaZyAFLiekChUFMyJdS8Lu3MavmziyqlRhIke+80HwKxpoZyzTC+NPGJJd5ky5ip0ufTK1KENzY/OabFTglroMxAOh3Gee4gm8UP0inr6T2b6RnqD0x8iUmT+S8LQplThEBQNX/hlXNGID8GY=","selinux_current_mode":"enforcing","rubyversion":"1.8.7","mtu_lo":"16436","selinux_config_policy":"targeted","uptime_hours":0,"selinux_config_mode":"enforcing","is_virtual":"false","ipaddress_eth0":"192.168.122.120","path":"/sbin:/usr/sbin:/bin:/usr/bin","operatingsystem":"CentOS","macaddress_eth0":"52:54:01:31:E4:79","kernelversion":"2.6.32","timezone":"EDT","augeasversion":"1.0.0","kernelrelease":"2.6.32-431.el6.x86_64","swapsize":"1023.99 MB","selinux_policyversion":"24","hostname":"proxybug","uptime_seconds":380,"facterversion":"1.6.18","osfamily":"RedHat","kernelmajversion":"2.6","ipaddress":"192.168.122.120","id":"foreman-proxy","domain":"zzz.lan","selinux":"true","interfaces":"eth0,lo","ipaddress_lo":"127.0.0.1","selinux_enforced":"true","architecture":"x86_64","selinux_mode":"targeted","network_lo":"127.0.0.0","swapfree":"1023.99 MB","operatingsystemrelease":"6.5","processor0":"Intel Core i7 9xx (Nehalem Class Core i7)","physicalprocessorcount":1,"fqdn":"proxybug.zzz.lan","netmask":"255.255.255.0","sshrsakey":"AAAAB3NzaC1yc2EAAAABIwAAAQEAoNFJSPPeOa365Pso8KIJDkkF4fmJJrIhHOYFA+OvKMVsu9cTZadAmxdky47zabVNkaNSRPp3Tyrk4Bxi2oKJRaiZ/c+LIrTDYNvYJ1DCqbxMHIKSFZvQBok/DJMghNKMP+q4SAX8SU01tBRgNms+fEmfQVKoie9fXo20WVWnJ3CbOYH+i1r8CRJXkFiccA0D9asWMV6HwgnGuzQ7tNdthMAioykI5hbWmmy9MHueZjY494wopjfObezkOfOrYwUrZnU9MoQFLpbb2L72r4MDTIxeRkgLyrgV4QZ2au5L7RBjH+YFiZ0ECgze/yR7j95yGtqjrSL2QovDIZU3SW1hrQ==","puppetversion":"2.7.25","memorytotal":"2.94 GB","hardwaremodel":"x86_64","macaddress":"52:54:01:31:E4:79","network_eth0":"192.168.122.0","memorysize":"2.94 GB","memoryfree":"2.42 GB","netmask_lo":"255.0.0.0","hardwareisa":"x86_64","uniqueid":"a8c0787a","processorcount":"1","vir100%[=====================================================================================>] 2,462       --.-K/s   in 0.04s   

2014-08-14 11:31:17 (60.3 KB/s) - written to stdout [2462/2462]
```

File: modules/root/root_api.rb [hunk 1]
```diff
@@ -4,7 +4,9 @@ class Proxy::RootApi < Sinatra::Base
 
   get "/features" do
     begin
-      @features = ::Proxy::Plugins.enabled_plugins.collect(&:plugin_name).sort - [:foreman_proxy]
+      plugin_names = ::Proxy::Plugins.enabled_plugins.collect(&:plugin_name).collect(&:to_s).sort
+
+      @features = plugin_names - [:foreman_proxy]
```
Comment (domcleal) [inline modules/root/root_api.rb]: foreman_proxy shows up as you're subtracting a symbol from an array of strings

File: lib/proxy/plugin.rb [hunk 2]
```diff
@@ -135,18 +135,16 @@ def settings
...
   def log_used_default_settings
-    unless settings.used_defaults.empty?
-      as_string = settings.defaults.select {|k,v| settings.used_defaults.include?(k)}.
-        sort.
-        collect {|c| ":#{c[0]}: #{c[1]}"}.
-        join(", ")
-      logger.info("'#{plugin_name}' settings were initialized with default values: #{as_string}")
-    end
+    settings.defaults.select {|k,v| settings.used_defaults.include?(k)}.
+      inject({}) {|acc, c| acc[c[0]] = c[1]; acc}.
```
Comment (domcleal) [inline lib/proxy/plugin.rb]: I think this is the only new line, but I can't see how it helps.  There's no string conversion happening.. so this still generates a hash of symbols to values:

```
[5] pry(#<Proxy::RootPlugin>)> settings.defaults.select {|k,v| settings.used_defaults.include?(k)}.inject({}) {|acc, c| acc[c[0]] = c[1]; acc}
=> {:enabled=>true}
```

Which will fail again when you call .sort on 140.

File: modules/root/root_api.rb (hunk 1 above)
Comment (lzap) [inline modules/root/root_api.rb]: Yeah, I was just about to write that. Will this work on 1.8?

Comment (domcleal) [inline modules/root/root_api.rb]: It should probably just be `plugin_names - 'foreman_proxy'`

Comment (dmitri-d) [inline modules/root/root_api.rb]: Nods; good catch, updated this and the test.

File: lib/proxy/plugin.rb (hunk 2 above)
Comment (dmitri-d) [inline lib/proxy/plugin.rb]: /me shakes his head. Nods, I meant to add a #to_s on the symbol (fixed); I wonder why the test didn't fail on 1.8.7 though.

**Comment by lzap** [comment]
Will test in the morning: http://koji.katello.org/koji/taskinfo?taskID=140579

**Comment by lzap** [comment]
Hmmm ok on RHEL6 but I don't see `/facts` url. Should I?

```
[root@proxybug ~]# rpm -Uvh --force "http://koji.katello.org/koji/getfile?taskID=140580&name=foreman-proxy-1.6.0-0.develop.el6.noarch.rpm"
Retrieving http://koji.katello.org/koji/getfile?taskID=140580&name=foreman-proxy-1.6.0-0.develop.el6.noarch.rpm
Preparing...                ########################################### [100%]
   1:foreman-proxy          ########################################### [100%]
[root@proxybug ~]# service foreman-proxy restart
Stopping foreman-proxy:                                    [  OK  ]
Starting foreman-proxy:                                    [  OK  ]
[root@proxybug ~]# wget https://localhost:8443/features --no-check-certificate --header "Accept: application/json" -qO -
["puppet","puppetca","tftp"]
```

Edit: I do have facter installed.

**Comment by domcleal** [comment]
The RPM scratch build above doesn't have the facts fix in commit 4bfe72540, in fact, I don't think it's even this branch as I can't see this fix in it.

**Comment by domcleal** [comment]
Merged as 7124635 for Foreman 1.6.0-RC2, thanks @witlessbird.

**Comment by lzap** [comment]
Thanks guys!

**Pull request closed without merging**